src/lib.rs

```diff
@@ -0,0 +1,111 @@
+extern crate tokio_trace;
+
+struct Span {
+    label: String,
+    ref_count: usize,
+}
+
+use std::fmt;
+use std::sync::Mutex;
+
+use tokio_trace::field::{Field, Visit};
+use tokio_trace::metadata::Level;
+use tokio_trace::{span, Event, Metadata, Subscriber};
+pub struct Visitor;
+
+impl Visit for Visitor {
+    fn record_debug(&mut self, field: &Field, value: &fmt::Debug) {
+        if field.name() == "message" {
+            print!("{:?}: ", value);
+        } else {
+            print!("{} = {:?} ", field.name(), value);
+        }
+    }
+}
+
+pub struct DemoSubscriber {
+    spans: Mutex<Vec<Span>>,
+    stack: Mutex<Vec<u64>>,
```

Comment:
Note that this is storing _one global stack_ of spans for the entire process. In `tokio-trace`, individual threads may be in separate spans, so if you want your subscriber to behave properly in a multithreaded context, you'll probably want this to be a thread-local rather than global.

Reply:
Oh that makes sense! Also no synchronization would be required for stack access.

1. For reference, a span id might move between threads (e.g. a future storing its span being moved to another thread/executor), but at that point it shouldn't be on the stack anymore? Since the span would be a member of the future, moving it to another thread while its span is on the stack should be blocked by the borrow checker.

2. But what about a thread pool? The tasks would loose the span stack of their outer caller and live within the span stack of the thread pool. I think that would be a good use case where you might want to set the parent of the task spans? Otherwise it would probably be hard to reason about why and where the work happend.

3. Actually while i'm thinking about it, this has affects the architecture heavily. The (out-of-process) console might have interest in the span stack for visualization purposes (tree style view, timelines), but also needs to know which threads are running and on which thread an event/record was issued. So the subscriber would need to assign and manage internal ids for [ThreadId](https://doc.rust-lang.org/std/thread/struct.ThreadId.html)s.

Reply:
These are great questions --- you definitely have the right intuitions here. 

>     1. For reference, a span id might move between threads (e.g. a future storing its span being moved to another thread/executor), but at that point it shouldn't be on the stack anymore? Since the span would be a member of the future, moving it to another thread while its span is on the stack should be blocked by the borrow checker.
> 

Not necessarily --- there isn't a rule stating that only a single thread can be executing in a span at a time. We might enter a span and then spawn a future on the threadpool which is _also_ in that span, for example:
```rust
let mut span = span!("foo");
let mut span2 = span.clone();
span.enter(move || {
    tokio::spawn(future.instrument(span2));
    // do other stuff inside the span...
})
``` 
I think that we should only consider a span to no longer be on the stack for a given thread when it informs us that it's been exited. 

>     2. But what about a thread pool? The tasks would loose the span stack of their outer caller and live within the span stack of the thread pool. I think that would be a good use case where you might want to set the parent of the task spans? Otherwise it would probably be hard to reason about why and where the work happend.
> 
You're entirely correct --- this is why it's necessary to track a concept of spans having _parents_ as well as having a span stack. In my experience, the main use of the stack is just for knowing what span we'll be in when the current span is exited, not for displaying a hierarchy of spans, we'd want to use parents for this.

>     3. Actually while i'm thinking about it, this has affects the architecture heavily. The (out-of-process) console might have interest in the span stack for visualization purposes (tree style view, timelines), but also needs to know which threads are running and on which thread an event/record was issued. So the subscriber would need to assign and manage internal ids for [ThreadId](https://doc.rust-lang.org/std/thread/struct.ThreadId.html)s.

When we've added `tokio-trace` instrumentation to the other tokio crates (soon!), each `tokio-threadpool` worker thread will do all its work inside of a span annotated with that worker's ID. So some of that work will be done for us...